_published/0004-core-lib-repo.md

```diff
@@ -0,0 +1,198 @@
+component's domains.
+
+Instead, it is a place of common utility code. What you will find in this
+utility repository is Go code that features:
+
+- Reading configuration from files and/or environment variables
+- Parsing `.wharf-ci.yml` files
+- Logging in a unified manner
+- Serving common endpoints such as `GET /version`
+
+Sample:
+
+```go
+package main
+
+import (
+    "io/ioutil"
+    _ "embed"
+
+    "github.com/iver-wharf/wharf-core/pkg/app"
+    "github.com/iver-wharf/wharf-core/pkg/config"
+    "github.com/iver-wharf/wharf-core/pkg/ginutils"
+    "github.com/iver-wharf/wharf-core/pkg/log"
+    "github.com/iver-wharf/wharf-core/pkg/wharfyml"
+
+    "github.com/gin-gonic/gin"
+)
+
+type DBConfig struct {
+    Host     string `yaml:"host" env:"HOST"`
+    Port     int    `yaml:"port" env:"PORT"`
+    Username string `yaml:"username" env:"USERNAME"`
+    Password string `yaml:"password" env:"PASSWORD"`
+}
+
+type Config struct {
+    Logging log.Config      `yaml:"logging" env:"LOGGING"`
+    API     ginutils.Config `yaml:"api" env:"API"`
+    DB      DBConfig        `yaml:"db" env:"DB"`
+}
+
+type Version struct {
+    app.Version
+}
+
+// go:embed version.yaml
+var versionFile []byte
+
+// @title Sample program
+// @description This program takes use of the utility repository to load in
+// @description the config and version of the app.
+func main() {
+    var version Version
+    config.UnmarshalYAML(versionFile, &version, config.Options{})
+
+    log.Infof("sample-program version=%s", version.AppVersion)
+
+    var config Config
+    configFile, _ := ioutil.ReadFile("config.yaml")
+    config.UnmarshalYAML(configFile, &config, config.Options{
+      AllowEnvironmentVariables: true,
+    })
+
+    log.SetConfig(config.Logging)
+
+    log.Info("Successfully read config.")
+
+    wharfDefFile, _ := ioutil.ReadFile(".wharf-ci.yml")
+    wharfDef, _ := wharfyml.ParseDefinition(wharfDefFile)
+    log.Infof("Read a .wharf-ci.yml file that contains %d stages.", len(wharfDef.Stages))
+
+    r := gin.Default()
+
+    // func ApplyConfig(engine *gin.Engine, ginutils.Config)
+    ginutils.ApplyConfig(r, config.API)
+
+    // func AddVersionEndpoint(engine *gin.Engine, version interface{})
+    ginutils.AddVersionEndpoint(r, version)
+
+    r.Run()
+}
+```
+
+```yaml
+# config.yaml
+logging:
+  level: debug
+  blacklist:
+    - gin
+    - gorm
+api:
+  allowCors: true
+
+db:
+  host: localhost
+  port: 5432
+  username: postgres
+  password: changeit # Can be overritten with WHARF_DB_PASSWORD environment variable
+```
+
+```yaml
+# version.yaml
+$schema: https://github.com/iver-wharf/wharf-core/raw/master/pkg/app/version-schema.json
+appVersion: v1.0.0
+buildGitCommit: 5971d3b585a722536730c39a22aa3148993f2985
+buildRef: 123
+buildDate: 2021-05-12T13:55:00+02:00
+```
+
+## Compatibility
+
+In architectural terms, this utility repository has to be "stable". Meaning it
+will be a hassle to update as there will be so many components relying on it.
```

Comment:
I mean, we shouldn't break things for the sake of it, but you can specify a certain version of the package when importing it right? So if you make a breaking change, it's not like it will all of a sudden break everything depending on it.

Reply:
Indeed. What I tried to lift here was the potential issue of for example finding a bug or making a bigger change that has to then be reflected in 4-5 other repos in 4-5 PRs.

But with the assumption of versioning the core lib (which is my intention either way) then this issue get decreased to "we'll change it in due time anyway, no need to rush"

Reply:
Resolving